src/main/java/com/xemantic/githubusers/logic/user/UserListPresenter.java

```diff
@@ -44,52 +44,53 @@
  */
 public class UserListPresenter extends Presenter {
 
-  private final Observable<UserQueryEvent> userQuery$;
+  private final UserListView view;
 
-  private final UserService userService;
-
-  private final Provider<UserView> userViewProvider;
-
-  private final Provider<UserPresenter> userPresenterProvider;
+  private final UserPresenterFactory userPresenterFactory;
 
   private final List<UserPresenter> displayedUserPresenters = new LinkedList<>();
 
   private final int userListPageSize;
 
   private final int gitHubUserSearchLimit;
 
-  private UserListView view;
-
   @Inject
   public UserListPresenter(
+      UserListView view,
       Observable<UserQueryEvent> userQuery$,
       UserService userService,
-      Provider<UserView> userViewProvider,
-      Provider<UserPresenter> userPresenterProvider,
+      UserPresenterFactory userPresenterFactory,
       @Named("userListPageSize") int userListPageSize,
       // "Only the first 1000 search results are available"
       @Named("gitHubUserSearchLimit") int gitHubUserSearchLimit
   ) {
-    this.userQuery$ = userQuery$;
-    this.userService = userService;
-    this.userViewProvider = userViewProvider;
-    this.userPresenterProvider = userPresenterProvider;
+
+    this.view = view;
+    this.userPresenterFactory = userPresenterFactory;
     this.userListPageSize = userListPageSize;
     this.gitHubUserSearchLimit = gitHubUserSearchLimit;
-  }
 
-  /**
-   * Starts and initializes the presenter with given {@code view}.
-   *
-   * @param view the view.
-   */
-  public void start(UserListView view) {
-    this.view = view;
-    on(userQuery$
+    register(userQuery$
         .filter(event -> ! event.getQuery().trim().isEmpty())  // kick out empty queries
         .map(UserQueryEvent::getQuery)
-        .switchMap(this::loadUsers)// will dispose stale subscriptions from previous requests
-    ).call(this::handlePage);
+        .switchMap(query -> { // switchMap will dispose stale subscriptions from previous request
+          AtomicInteger pager = new AtomicInteger(1);
```

Comment:
iuff, you continue creating too many side effects. Sorry to be repetitive, hehe but I really think this is not going to end well both you and anyone who will see this project will really have a bad feeling about Observable if you do that. Observables are super-descriptive and looks super simple, but it really does a lot of things and it if you do not try to make it a bit more functional and side-effect free it will really be crazy when bugs start appearing. Debugging is not easy at all, but if you keep it side-effect free then it is easy to test and difficult to introduce bugs.

There are many tutorials about reactivex and side-effect like http://www.introtorx.com/content/v1.0.10621.0/09_SideEffects.html, also you need to learn how to use the scan operator which is the way of having state in an observable without side-effect neither mutations.

I promise that hehe, if you start programming with observables that way you are going to regret, and maybe you end up even thinking that observables are the devil, and on the other hand, if you try to apply the basic rules of NEVER subscribe, never ever subscribe and observable inside an observable and never use external state like this atomic int you are going to force yourself to actually learn a bit about reactive functional programming which is going to make your code MUCH MUCH more predictable.

This has examples and various interesting links...
https://www.learnrxjs.io/operators/transformation/scan.html

Reply:
It would be so easy if I didn't have to populate the first page on new query. :( I struggle finding easy solution for this problem. For this reason I introduced `Trigger.oneTime()` which doesn't feel right.

Reply:
I'll try! need some time, I hope this night I'll do it! Anyway, the most important is that "it doesn't feel right" hehe so is just a bit of time that you end up fixing it 🙌 ... to be honest, I do those hacks too! hehe but I really try to fix or hide as much as possible, as this is a showcase open-sourced project, all those hacks should really not end up exposed as those are really bad practices (although necessary bc reactive-functional has a hard learning curve as GWT itself 😜).